src/reduce.py

```diff
@@ -12,8 +12,13 @@ def reduce(f: Callable[[A], B], x: list[A]) -> B:
     >>> reduce(lambda x,y: x+y, [1, 2, 3])
     6
     """
-    assert len(x) >= 2
-    ...  # FIXME
+    if len(x) <= 2:
+        return x[0]
```

Comment:
If `x = [1, 2]` (with `len(x) == 2`) you want to call `f` with `f(x[0],x[1])`, not return `x[0]`. On the other hand, if `len(x) == 0` (which is also `len(x) <= 2`), getting `x[0]` will crash.

To make this work, you probably want to assert that `if 0 < len(x) < 2` (or just `if len(x) == 1`) then `return x[0]`.

This isn't always such an obvious choice, though.

The input list is a list of type `A`, and you must return something of type `B`. If `A != B`, `x[0]` has the wrong type. Now, this isn't the situation we are in here because we assume that `f` can take an `A` as its second argument (because we start out by applying `f(x[0],x[1])`), so whatever `A` is must also be a `B`. In a more general setting, where the two types are not compatible, you need an initial value for `f` so it goes

```python
def reduce(f: Callable[[A,B], B], x: list[A], init: B) -> B:
    match len(x):
       case 0: return init
       case 1: return f(x[0], init)
       case _:
           # handle the general case
           ...
```

Reply:
Thanks for the explanation!